### ISSUE TITLE
Karma is deprecated

### ISSUE THREAD

**Issue by jonkoops** [Leaflet/Leaflet#8939]
### Checklist

- [X] I've searched through the [plugins](https://leafletjs.com/plugins.html) to make sure this feature isn't already available, or think it shouldn't require a plugin.
- [X] I've searched through the [current issues](https://github.com/Leaflet/Leaflet/issues) to make sure this feature hasn't been requested already.
- [X] I agree to follow the [Code of Conduct](https://github.com/Leaflet/Leaflet/blob/main/CODE_OF_CONDUCT.md) that this project adheres to.

### Motivation

Karma [has been deprecated](https://github.com/karma-runner/karma#karma-is-deprecated-and-is-not-accepting-new-features-or-general-bug-fixes) today, so we'll likely have to migrate away from it at some point.

### Suggested solution

We might want to consider [Web Test Runner](https://modern-web.dev/docs/test-runner/overview/) as the [blog post](https://blog.angular.io/moving-angular-cli-to-jest-and-web-test-runner-ef85ef69ceca) from the Angular team suggests.

**Comment by mourner** [comment]
I thought we would migrate away from Karma sooner or later anyway — it's a very old project and a pretty stale one. Web Test Runner looks pretty nice.

**Comment by jonkoops** [comment]
@mourner I wonder if the performance issues that we're experiencing under #8936 might also be resolved migrating away from Karma. I think I will take a look at how much effort it might be to move away from Karma first.

**Comment by jonkoops** [comment]
@Falke-Design as part of our migration to Web Test Runner we need all of our dependencies used during the testing process to be ESM compatible. Since you are maintaining `ui-event-simulator` could you review [the PR](https://github.com/Falke-Design/ui-event-simulator/pull/2) implementing the aforementioned there?

**Comment by Falke-Design** [comment]
@jonkoops done

**Comment by jonkoops** [comment]
@Falke-Design  Awesome thanks! Could you also publish a 2.0 for that package?

**Comment by Falke-Design** [comment]
Should be already there 🤔

**Comment by jonkoops** [comment]
Strange, it is not showing up [on NPM](https://www.npmjs.com/package/ui-event-simulator?activeTab=versions).

**Comment by Falke-Design** [comment]
I forgot to publish ... 😄 I will do it at the evening (in 6h)

**Comment by jonkoops** [comment]
I also had to refactor `prosthetic-hand` to a pure ESM package so it can be consumed by Web Test Runner. @IvanSanchez can I ask you to review this PR and publish a new version? https://github.com/Leaflet/prosthetic-hand/pull/25

**Comment by Falke-Design** [comment]
> I forgot to publish ... 😄 I will do it at the evening (in 6h)

@jonkoops done

**Comment by mourner** [comment]
Tried porting a small part of the tests to Web Test Runner to get a feel for it — looks like it works pretty well overall. Here's the commit: https://github.com/Leaflet/Leaflet/commit/e9a3cf3878e7e18bc1822a8cf9c37d17866e1b48

A few hiccups I've encountered:

- We'll probably have to mock `package.json` because browsers don't support importing JSON like it was JS (or maybe `with` / `assert` keywords would help, not sure about browser support though).
- `GeneralSpec.js` can't test global variables because the test will import Leaflet as a module, so it won't get the global `L`.

I guess we could address both by setting the built Leaflet version as the entry point in the import maps rather than `src/Leaflet.js`, but not sure — didn't we want the source to be a viable module entry point for v2? In that case it would defeat the purpose, right? @jonkoops

**Comment by jonkoops** [comment]
> We'll probably have to mock `package.json` because browsers don't support importing JSON like it was JS (or maybe `with` / `assert` keywords would help, not sure about browser support though).

Yeah, basically only Chrome supports this at the moment, and it only supports the now defunct `assert` keyword. We could work around this for now by handling this at compile-time, or perhaps by fetching the JSON.

> `GeneralSpec.js` can't test global variables because the test will import Leaflet as a module, so it won't get the global `L`.

Ah yes, I hadn't considered that. We'll have to find a way to get that working.

> I guess we could address both by setting the built Leaflet version as the entry point in the import maps rather than `src/Leaflet.js`, but not sure — didn't we want the source to be a viable module entry point for v2? In that case it would defeat the purpose, right? @jonkoops

Yeah, ideally we'd be able to load it directly from source, but there are likely still benefits to having a bundled version. We'll have to evaluate this based on user demand I think.

I think right now the most critical thing is that we want to move the test suite away from Karma, we can worry about details for the distribution later. We might also consider using [Vitest](https://vitest.dev/), which has the advantage that we can rely on Vite being the bundler, of which my experience has been pretty good. It also has a built-in [browser mode](https://vitest.dev/guide/browser.html).

@mourner perhaps you could raise the Web Test Runner code as a draft PR? We can see if we can do some collaboration to get it landed.

**Comment by mourner** [comment]
@jonkoops started a draft PR here: https://github.com/Leaflet/Leaflet/pull/9242

I also tried Vitest but somehow found it much clunkier and more difficult to get working, at least for this codebase; perhaps this is due to its browser mode still being experimental and not mature enough.

**Comment by jonkoops** [comment]
Yeah, just wanted to put it out there as an option. I think Web Test Runner might make more sense for us now, also because it will allow us to keep the test suite pretty much unmodified otherwise.

**Comment by jonkoops** [comment]
Alternatively we might also consider [Uvu](https://github.com/lukeed/uvu), but it seems extremely minimal.

**Comment by mourner** [comment]
@jonkoops love uvu, would be really awesome if we could get it to work with multiple browsers on CI and enough flexibility. The minimalistic approach fits the one from Leafet really well, and the new bloated frameworks are really tiring (very hard to debug and maintain).

**Comment by jonkoops** [comment]
Yeah, totally agree with that assessment.

**Comment by mourner** [comment]
I'll take a stab a this again — we need to migrate to something even if it's not perfect. Staying with Karma / Mocha is too painful.
